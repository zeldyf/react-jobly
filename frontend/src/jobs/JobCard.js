import React, { useState, useEffect, useContext } from 'react';
import { Card, CardBody, CardTitle, CardText, Button } from 'reactstrap';

import UserContext from '../UserContext';
import './JobCard.css';

/** Individual card component for JobList.
 *
 * Show simple info about a job on the list
 *
 * Is rendered by JobCardList to show a "card" for each job.
 *
 * Receives apply func prop from parent, which is called on apply.
 *
 * JobCardList -> JobCard
 */

const JobCard = ({ id, title, salary, equity, companyName }) => {
  const { hasAppliedToJob, applyToJob } = useContext(UserContext);
  const [applied, setApplied] = useState();

  useEffect(() => {
    console.debug('JobCard useEffect update applied status', 'id=', id);
    setApplied(hasAppliedToJob(id));
  }, [id, hasAppliedToJob]);

  async function handleApply(evt) {
    if (hasAppliedToJob(id)) return;
    applyToJob(id);
    setApplied(true);
  }

  return (
    <Card key={id} className="JobCard mb-3">
      <CardBody>
        <CardTitle>{title}</CardTitle>
        <h6>{companyName}</h6>
        <CardText>Salary: {salary ? salary : 'TBA'}</CardText>
        <CardText>Equity: {equity ? equity : 0}</CardText>
        <Button
          onClick={handleApply}
          disabled={applied}
          color="danger"
          className="font-weight-bold text-uppercase float-right"
        >
          {applied ? 'Applied' : 'Apply'}
        </Button>
      </CardBody>
    </Card>
  );
};

export default JobCard;
