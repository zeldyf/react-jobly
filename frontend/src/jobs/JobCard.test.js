import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import JobCard from './JobCard';
import UserContext from '../UserContext';

const job = {
  id: 1,
  title: 'Software Engineer',
  salary: 100000,
  equity: 0.1,
  companyName: 'Example Co.',
};

const applyToJob = jest.fn();

const hasAppliedToJob = jest.fn(() => false);

const userContextValues = {
  hasAppliedToJob,
  applyToJob,
};

test('renders without crashing', () => {
  render(
    <UserContext.Provider value={userContextValues}>
      <JobCard {...job} />
    </UserContext.Provider>
  );
});

test('matches snapshot', () => {
  const { asFragment } = render(
    <UserContext.Provider value={userContextValues}>
      <JobCard {...job} />
    </UserContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('applies to job', () => {
  const { getByText } = render(
    <UserContext.Provider value={userContextValues}>
      <JobCard {...job} />
    </UserContext.Provider>
  );

  const applyButton = getByText('Apply');
  fireEvent.click(applyButton);
  expect(applyToJob).toHaveBeenCalledTimes(1);
  expect(applyToJob).toHaveBeenCalledWith(job.id);
});

