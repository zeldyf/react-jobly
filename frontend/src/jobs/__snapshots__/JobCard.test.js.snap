// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="JobCard mb-3 card"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title"
      >
        Software Engineer
      </h5>
      <h6>
        Example Co.
      </h6>
      <p
        class="card-text"
      >
        Salary: 100000
      </p>
      <p
        class="card-text"
      >
        Equity: 0.1
      </p>
      <button
        class="font-weight-bold text-uppercase float-right btn btn-danger"
        type="button"
      >
        Apply
      </button>
    </div>
  </div>
</DocumentFragment>
`;
